### DISCUSSION TITLE
Support for OpenAPI JSON Schema type attribute as array 🙏

### DISCUSSION THREAD

**Discussion by autoluminescent** [fastapi/fastapi#12787] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from pydantic import BaseModel, ConfigDict
from fastapi import FastAPI

def _remove_any_of(schema):
    # this is just an example, obviously in real code we'd take the generated json schema, look for all the `anyOf` instances and replace them
    schema['properties']['foo'].pop("anyOf")
    schema['properties']['foo']['type'] = ['integer', 'null']

class MyModel(BaseModel):
    model_config = ConfigDict(json_schema_extra=_remove_any_of)
    foo: int | None

app = FastAPI()

@app.post("/foo")
def post_foo(model: MyModel) -> None:
    ...

print(app.openapi())
```

### Description

* Copy/paste example to REPL
* I expect an OpenAPI JSON to be rendered
* An error is raised instead, because the OpenAPI schema defined by FastAPI does not support an array in the [`type` field](https://github.com/fastapi/fastapi/blob/master/fastapi/openapi/models.py#L160).

In our case, we often need to flatten occurrences of `anyOf` (where possible) in order facilitate the processing of these JSON schemas in other parts of the code. Since they are valid JSON schemas, it would be nice if FastAPI supports them.

### Operating System

macOS

### Operating System Details

Sequoia 15.1

### FastAPI Version

0.115.4

### Pydantic Version

2.8.2

### Python Version

3.11.10

**Comment by Kfir-G** [comment]
@autoluminescent  FastAPI doesn’t support using an array for the `type` field in OpenAPI JSON Schema (like `["integer", "null"]`). Instead, it uses `anyOf` or `oneOf` to handle optional types, which matches OpenAPI 3.0 specs but not the JSON Schema standard.

**Quick Fixes:**
1. Stick with `anyOf` for now, since it's supported.
2. Customize the OpenAPI schema with a function to modify it after FastAPI generates it:
   ```python
   from fastapi.openapi.utils import get_openapi

   def custom_openapi():
       if app.openapi_schema:
           return app.openapi_schema
       openapi_schema = get_openapi(
           title="My API",
           version="1.0.0",
           description="API with modified schema",
           routes=app.routes,
       )
       # Make your adjustments here if needed
       app.openapi_schema = openapi_schema
       return app.openapi_schema

   app.openapi = custom_openapi
   ```
3. Request support for OpenAPI 3.1 in FastAPI, which aligns better with JSON Schema.

This will let you handle schema quirks without breaking things

**Comment by YuriiMotov** [comment]
This is not supported by openapi specification.
If you think I'm wrong, please re-open the discussion and point to the place in the specification where it's allowed

**Reply by autoluminescent** [reply]
Hi Yurii, thanks for the reply. It's ok, I don't really need this anymore, so the issue can remain closed. But I was curious about where I got this idea in the first place, and I found out that [JSON Schema](https://json-schema.org/understanding-json-schema/reference/type) does support this. Also in the [draft](https://json-schema.org/draft/2020-12/json-schema-core) referenced by the OpenAPI 3.1.0 spec, there are a couple of examples with arrays of types. However, I do believe that OpenAPI 3.0 [does not support this](https://swagger.io/docs/specification/v3_0/data-models/data-types/#type) indeed.
